File: reference-architectures/app/api/typescript/openai-agents-sdk/src/activity-routes.ts
import type { FastifyInstance, FastifyReply, FastifyRequest } from 'fastify';
import type {
  ActivityResolution,
  Conversation,
  ConversationDetail,
  ConversationList,
  HealthResponse,
  Message
} from './types.ts';

type ActivityMode = 'discovery' | 'planning' | 'staffing';
type ActivityStatus = 'active' | 'resolved';

interface AgentRuntime {
  createConversation(metadata?: Record<string, unknown>): Promise<Conversation>;
  listConversations(offset?: number, limit?: number): Promise<ConversationList>;
  getConversation(conversationId: string): Promise<ConversationDetail | undefined>;
  sendMessage(conversationId: string, content: string): Promise<Message | undefined>;
  sendMessageStream(conversationId: string, content: string, onChunk: (chunk: string) => void): Promise<void>;
  checkHealth(): Promise<HealthResponse>;
}

interface ActivityRecord {
  mode: ActivityMode;
  status: ActivityStatus;
  outcome?: ActivityResolution;
}

const store = new Map<string, ActivityRecord>();

const syntheticMessages: Record<ActivityMode, string> = {
  discovery:
    'I am qualifying a new customer opportunity. Lead a short discovery conversation, ask targeted questions, and conclude with a concise qualification summary.',
  planning:
    'I need an account plan for an active customer. Guide me through priorities, risks, and next steps, then conclude with a concise planning summary.',
  staffing:
    'I need to staff an account team for a customer engagement. Interview me for the needed context and conclude with a clear staffing recommendation.'
};

function errorReply(reply: FastifyReply, status: number, code: string, message: string): void {
  void reply.status(status).send({ code, message });
}

function modeFrom(metadata: Record<string, unknown> | undefined): ActivityMode {
  const mode = metadata?.['mode'];
  return mode === 'planning' || mode === 'staffing' || mode === 'discovery' ? mode : 'discovery';
}

function activityConversation(conversation: Conversation) {
  const record = store.get(conversation.id);
  return {
    id: conversation.id,
    mode: record?.mode ?? modeFrom(conversation.metadata),
    status: record?.status ?? 'active',
    ...(record?.outcome ? { outcome: record.outcome } : {}),
    createdAt: conversation.createdAt,
    lastMessageAt: conversation.updatedAt,
    messageCount: 0
  };
}

function activityDetail(detail: ConversationDetail) {
  const record = store.get(detail.id);
  return {
    ...activityConversation(detail),
    messageCount: detail.messages.length,
    messages: detail.messages
  };
}

function markResolved(conversationId: string, outcome: ActivityResolution | undefined): void {
  if (!outcome) return;
  const record = store.get(conversationId);
  if (record) {
    record.status = 'resolved';
    record.outcome = outcome;
  }
}

function captureResolvedEvent(chunk: string): ActivityResolution | undefined {
  const eventIndex = chunk.indexOf('event: activity.resolved');
  if (eventIndex < 0) return undefined;
  const dataLine = chunk.slice(eventIndex).split(/\r?\n/).find((line) => line.startsWith('data: '));
  if (!dataLine) return undefined;
  return JSON.parse(dataLine.slice('data: '.length)) as ActivityResolution;
}

export function registerActivityRoutes(app: FastifyInstance, runtime: AgentRuntime, modelName: string): void {
  async function start(mode: ActivityMode, req: FastifyRequest, reply: FastifyReply): Promise<void> {
    try {
      const conversation = await runtime.createConversation({ mode });
      store.set(conversation.id, { mode, status: 'active' });
      await reply.status(201).send({
        id: conversation.id,
        mode,
        status: 'active',
        syntheticMessage: syntheticMessages[mode],
        createdAt: conversation.createdAt
      });
    } catch (err) {
      req.log.error({ err, mode }, 'Failed to start activity conversation');
      errorReply(reply, 500, 'agent_error', 'Failed to start activity conversation');
    }
  }

  app.get('/health/deep', { logLevel: 'silent' }, async (_req, reply) => {
    const startTime = Date.now();
    const health = await runtime.checkHealth();
    const healthy = health.status === 'healthy';
    await reply.status(healthy ? 200 : 503).send({
      status: healthy ? 'healthy' : 'degraded',
      dependencies: [{ name: 'agent-runtime', status: health.status, latencyMs: Date.now() - startTime }]
    });
  });

  app.post('/api/activities/discovery', async (req, reply) => start('discovery', req, reply));
  app.post('/api/activities/planning', async (req, reply) => start('planning', req, reply));
  app.post('/api/activities/staffing', async (req, reply) => start('staffing', req, reply));

  app.get('/api/activities/conversations', async (req, reply) => {
    const query = req.query as Record<string, string | undefined>;
    const offset = query['offset'] ? Number.parseInt(query['offset'], 10) : 0;
    const limit = query['limit'] ? Number.parseInt(query['limit'], 10) : 20;
    const list = await runtime.listConversations(offset, limit);
    await reply.send({
      conversations: list.items.map(activityConversation),
      offset: list.offset,
      limit: list.limit,
      total: list.total
    });
  });

  app.get('/api/activities/conversations/:conversationId', async (req, reply) => {
    const { conversationId } = req.params as { conversationId: string };
    const detail = await runtime.getConversation(conversationId);
    if (!detail) {
      errorReply(reply, 404, 'not_found', `Conversation ${conversationId} not found`);
      return;
    }
    await reply.send(activityDetail(detail));
  });

  app.post('/api/activities/conversations/:conversationId/messages', async (req, reply) => {
    const { conversationId } = req.params as { conversationId: string };
    const message = (req.body as { message?: unknown } | undefined)?.message;
    if (typeof message !== 'string' || message.length === 0) {
      errorReply(reply, 400, 'bad_request', 'Missing required field: message');
      return;
    }

    if (String(req.headers['accept'] ?? '').includes('text/event-stream')) {
      void reply.hijack();
      reply.raw.writeHead(200, {
        'Content-Type': 'text/event-stream',
        'Cache-Control': 'no-cache',
        Connection: 'keep-alive'
      });

      let captured: ActivityResolution | undefined;
      try {
        await runtime.sendMessageStream(conversationId, message, (chunk) => {
          captured ??= captureResolvedEvent(chunk);
          reply.raw.write(chunk);
        });
        markResolved(conversationId, captured);
      } catch (err) {
        req.log.error({ err, conversationId }, 'Failed to stream activity message');
        reply.raw.write(
          `event: error\ndata: ${JSON.stringify({ code: 'agent_error', message: 'Failed to stream message' })}\n\n`
        );
      } finally {
        reply.raw.end();
      }
      return;
    }

    const result = await runtime.sendMessage(conversationId, message);
    if (!result) {
      errorReply(reply, 404, 'not_found', `Conversation ${conversationId} not found`);
      return;
    }
    markResolved(conversationId, result.resolution);
    await reply.send(result);
  });

  app.get('/api/activities/stats', async (_req, reply) => {
    const list = await runtime.listConversations(0, 100);
    const byMode = {
      discovery: { total: 0, active: 0, resolved: 0 },
      planning: { total: 0, active: 0, resolved: 0 },
      staffing: { total: 0, active: 0, resolved: 0 }
    };
    for (const conversation of list.items) {
      const record = store.get(conversation.id);
      const mode = record?.mode ?? modeFrom(conversation.metadata);
      const status = record?.status ?? 'active';
      byMode[mode].total += 1;
      byMode[mode][status] += 1;
    }
    const resolvedConversations = byMode.discovery.resolved + byMode.planning.resolved + byMode.staffing.resolved;
    await reply.send({
      totalConversations: list.items.length,
      activeConversations: list.items.length - resolvedConversations,
      resolvedConversations,
      byMode
    });
  });

  app.post('/chat', async (req, reply) => {
    const body = req.body as { message?: unknown; conversationId?: unknown } | undefined;
    const message = body?.message;
    if (typeof message !== 'string' || message.length === 0) {
      errorReply(reply, 400, 'bad_request', 'Missing required field: message');
      return;
    }
    const conversationId =
      typeof body?.conversationId === 'string' && body.conversationId.length > 0
        ? body.conversationId
        : (await runtime.createConversation({ mode: 'discovery', chat: true })).id;
    const response = await runtime.sendMessage(conversationId, message);
    await reply.send({ conversationId, reply: response?.content ?? '', model: modelName });
  });
}

File: reference-architectures/app/api/csharp/microsoft-agent-framework/ActivityRoutes.cs
using System.Collections.Concurrent;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace CairaAgent;

public static class ActivityRoutes
{
    private sealed class ActivityRecord
    {
        public required string Mode { get; init; }
        public string Status { get; set; } = "active";
        public ActivityOutcome? Outcome { get; set; }
    }

    private static readonly ConcurrentDictionary<string, ActivityRecord> Store = new();

    private static readonly Dictionary<string, string> SyntheticMessages = new()
    {
        ["discovery"] = "I am qualifying a new customer opportunity. Lead a short discovery conversation, ask targeted questions, and conclude with a concise qualification summary.",
        ["planning"] = "I need an account plan for an active customer. Guide me through priorities, risks, and next steps, then conclude with a concise planning summary.",
        ["staffing"] = "I need to staff an account team for a customer engagement. Interview me for the needed context and conclude with a clear staffing recommendation.",
    };

    public static void MapRoutes(WebApplication app, ConversationStore conversations, WorkflowRunner runner)
    {
        app.MapGet("/health/deep", () =>
        {
            var health = runner.CheckHealth();
            var ok = health.Status == "healthy";
            return ok
                ? Results.Ok(new ApiHealthResponse("healthy", [new DependencyHealth("agent-runtime", health.Status)]))
                : Results.Json(new ApiHealthResponse("degraded", [new DependencyHealth("agent-runtime", health.Status)]), statusCode: 503);
        });

        app.MapPost("/api/activities/discovery", () => Start("discovery", conversations));
        app.MapPost("/api/activities/planning", () => Start("planning", conversations));
        app.MapPost("/api/activities/staffing", () => Start("staffing", conversations));

        app.MapGet("/api/activities/conversations", (int? offset, int? limit) =>
        {
            var list = conversations.List(offset ?? 0, limit ?? 20);
            return Results.Ok(new ActivityConversationList(
                list.Items.Select(ToActivityConversation).ToList(),
                list.Offset,
                list.Limit,
                list.Total));
        });

        app.MapGet("/api/activities/conversations/{conversationId}", (string conversationId) =>
        {
            var detail = conversations.Get(conversationId);
            return detail == null
                ? Results.Json(new ErrorResponse("not_found", $"Conversation {conversationId} not found"), statusCode: 404)
                : Results.Ok(ToActivityConversationDetail(detail));
        });

        app.MapPost("/api/activities/conversations/{conversationId}/messages", async (
            string conversationId,
            HttpContext context) =>
        {
            var body = await context.Request.ReadFromJsonAsync<ActivityMessageRequest>();
            if (string.IsNullOrWhiteSpace(body?.Message))
            {
                return Results.Json(new ErrorResponse("bad_request", "Missing required field: message"), statusCode: 400);
            }

            if (context.Request.Headers.Accept.ToString().Contains("text/event-stream"))
            {
                return Results.Stream(async stream =>
                {
                    ActivityOutcome? captured = null;
                    await runner.SendMessageStreamAsync(conversationId, body.Message, async chunk =>
                    {
                        captured ??= TryCaptureOutcome(chunk);
                        await stream.WriteAsync(Encoding.UTF8.GetBytes(chunk));
                        await stream.FlushAsync();
                    });
                    MarkResolved(conversationId, captured);
                }, contentType: "text/event-stream");
            }

            var message = await runner.SendMessageAsync(conversationId, body.Message);
            if (message == null)
            {
                return Results.Json(new ErrorResponse("not_found", $"Conversation {conversationId} not found"), statusCode: 404);
            }

            MarkResolved(conversationId, message.Resolution == null
                ? null
                : new ActivityOutcome(message.Resolution.Tool, message.Resolution.Result));
            return Results.Ok(message);
        });

        app.MapGet("/api/activities/stats", () =>
        {
            var list = conversations.List(0, 100);
            var modes = new Dictionary<string, ModeStats>
            {
                ["discovery"] = new(0, 0, 0),
                ["planning"] = new(0, 0, 0),
                ["staffing"] = new(0, 0, 0),
            };

            foreach (var conversation in list.Items)
            {
                var record = Store.GetValueOrDefault(conversation.Id);
                var mode = record?.Mode ?? ModeFrom(conversation.Metadata);
                var current = modes[mode];
                modes[mode] = record?.Status == "resolved"
                    ? current with { Total = current.Total + 1, Resolved = current.Resolved + 1 }
                    : current with { Total = current.Total + 1, Active = current.Active + 1 };
            }

            var resolved = modes.Values.Sum(m => m.Resolved);
            return Results.Ok(new ActivityStats(list.Items.Count, list.Items.Count - resolved, resolved, modes));
        });

        app.MapPost("/chat", async (ActivityMessageRequest request) =>
        {
            var conversation = conversations.Create(new() { ["mode"] = "discovery", ["chat"] = true });
            var message = await runner.SendMessageAsync(conversation.Id, request.Message);
            return Results.Ok(new ChatResponse(conversation.Id, message?.Content ?? "", "microsoft-agent-framework"));
        });
    }

    private static IResult Start(string mode, ConversationStore conversations)
    {
        var conversation = conversations.Create(new() { ["mode"] = mode });
        Store[conversation.Id] = new ActivityRecord { Mode = mode };
        return Results.Json(new ActivityStarted(conversation.Id, mode, "active", SyntheticMessages[mode], conversation.CreatedAt), statusCode: 201);
    }

    private static ActivityConversation ToActivityConversation(Conversation conversation)
    {
        var record = Store.GetValueOrDefault(conversation.Id);
        return new ActivityConversation(
            conversation.Id,
            record?.Mode ?? ModeFrom(conversation.Metadata),
            record?.Status ?? "active",
            conversation.CreatedAt,
            conversation.UpdatedAt,
            0,
            record?.Outcome);
    }

    private static ActivityConversationDetail ToActivityConversationDetail(ConversationDetail detail)
    {
        var baseConversation = ToActivityConversation(new Conversation(detail.Id, detail.CreatedAt, detail.UpdatedAt, detail.Metadata));
        return new ActivityConversationDetail(
            baseConversation.Id,
            baseConversation.Mode,
            baseConversation.Status,
            baseConversation.CreatedAt,
            baseConversation.LastMessageAt,
            detail.Messages.Count,
            detail.Messages,
            baseConversation.Outcome);
    }

    private static string ModeFrom(Dictionary<string, object>? metadata)
    {
        if (metadata != null && metadata.TryGetValue("mode", out var mode) && mode?.ToString() is "planning" or "staffing" or "discovery")
        {
            return mode.ToString()!;
        }
        return "discovery";
    }

    private static void MarkResolved(string conversationId, ActivityOutcome? outcome)
    {
        if (outcome == null || !Store.TryGetValue(conversationId, out var record)) return;
        record.Status = "resolved";
        record.Outcome = outcome;
    }

    private static ActivityOutcome? TryCaptureOutcome(string chunk)
    {
        if (!chunk.Contains("event: activity.resolved", StringComparison.Ordinal)) return null;
        var line = chunk.Split('\n').FirstOrDefault(l => l.StartsWith("data: ", StringComparison.Ordinal));
        if (line == null) return null;
        return JsonSerializer.Deserialize<ActivityOutcome>(line["data: ".Length..]);
    }
}

public sealed record ActivityMessageRequest([property: JsonPropertyName("message")] string Message);
public sealed record ActivityStarted(string Id, string Mode, string Status, string SyntheticMessage, string CreatedAt);
public sealed record ActivityOutcome(string Tool, Dictionary<string, object> Result);
public sealed record ActivityConversation(string Id, string Mode, string Status, string CreatedAt, string LastMessageAt, int MessageCount, ActivityOutcome? Outcome = null);
public sealed record ActivityConversationDetail(string Id, string Mode, string Status, string CreatedAt, string LastMessageAt, int MessageCount, List<Message> Messages, ActivityOutcome? Outcome = null);
public sealed record ActivityConversationList(List<ActivityConversation> Conversations, int Offset, int Limit, int Total);
public sealed record ModeStats(int Total, int Active, int Resolved);
public sealed record ActivityStats(int TotalConversations, int ActiveConversations, int ResolvedConversations, Dictionary<string, ModeStats> ByMode);
public sealed record DependencyHealth(string Name, string Status, int? LatencyMs = null);
public sealed record ApiHealthResponse(string Status, List<DependencyHealth> Dependencies);
public sealed record ChatResponse(string ConversationId, string Reply, string Model);
